### PR TITLE
Assignment5+6

### PR DESCRIPTION
**Pull request by dhoxie**
Partner: @xiokd

### CODE REVIEW

**Comment by MikePHaynes** [comment]
Will review!

**Review by MikePHaynes — COMMENTED** [review]
You guys seem to be on the right track here. It looks like you have begun to implement all of the things you have missing, I did not see a unit test for the "People" property though. I did get a few messages from the compiler as well.

1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔

   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔

   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of Spokane-based addresses. ❌
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔

   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, Zip. (Sort the addresses first then select). ✔
   - Be sure that `Person.Address` is also populated. ✔
   - Adding null validation to all the `Person` and `Address` properties is **optional**.
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ❌

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔

   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use the `people` parameter from `ISampleData.People` property for your data source. ✔
   - At a minimum, use `System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔
   - It is recommended that, at a minimum, you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result.

7. Given the implementation of `Node` in Assignment5

- Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔
- Add an `IEnumberable<T> ChildItems(int maximum)` method to `Node` that returns the remaining items with a maximum number of items returned less than `maximum`.  

## Extra Credit

- Implement the homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `IAsyncSampleData`). Refactor your `SampleData` and `SampleDataAsync` classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

## Fundamentals

- Ensure you enable:
  - nullable reference types is enabled ✔
  - net6 targeted ✔
  - C# 10.0 ✔
  - and enabled .NET analyzers for both projects ✔
- For this assignment, favor using Assert.AreEqual<T>() (the generic version) ✔
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔

File: Assignment/Assignment.Tests/SampleDataTest.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+            int result = sampleData.CsvRows.Count();
+
+            Assert.AreEqual(50, result);
+        }
+
+        [TestMethod]
+        public void GetUniqueSortedListOfStatesGivenCsvRows_UsingHardcodedList_Success()
+        {
+            SampleData sampleData = new SampleData();
```
Comment (MikePHaynes) [inline Assignment/Assignment.Tests/SampleDataTest.cs:26]: This line can be simplified.

File: Assignment/Assignment.Tests/SampleDataTest.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            // IEnumerable<Address> sampleDataAddresses = 
+            // remove non-Spokane addresses from list
+
+            // check Spokane addresses are in list
+        }
+
+        [TestMethod]
+        public void GetUniqueSortedListOfStatesGivenCsvRows_IsDistinct()
+        {
+            SampleData sampleData = new SampleData();
```
Comment (MikePHaynes) [inline Assignment/Assignment.Tests/SampleDataTest.cs:41]: This line can also be simplified.

File: Assignment/Assignment/Node.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return (IEnumerator) GetEnumerator();
+        }
+
+        public IEnumerable<T> ChildItems(int maximum)
+        {
+            if (maximum < 0) throw new ArgumentOutOfRangeException("Maximum must be greater than 0");
```
Comment (MikePHaynes) [inline Assignment/Assignment/Node.cs:72]: You should replace the "Maximum must be greater than 0" argument with one of the methods parameter names (nameof(maximum)).

**Comment by A-Tanner** [comment]
I will review this

File: Assignment/Assignment/Node.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+            return (IEnumerable<T>)ChildItems(maximum);
```
Comment (A-Tanner) [inline Assignment/Assignment/Node.cs:74]: This method creates an infinite loop!

**Review by A-Tanner — CHANGES_REQUESTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of Spokane-based addresses. ❌✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use System.Linq.Enumerable.Aggregate LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.
❌ -->Method currently creates an infinite loop by calling itself with the same parameter
Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌
Fundamentals
Ensure you enable:
nullable reference types is enabled ✔
net6 targeted ✔
C# 10.0 ✔
and enabled .NET analyzers for both projects ✔
For this assignment, favor using Assert.AreEqual() (the generic version) ❌✔
All of the above should be unit tested ❌✔
Choose simplicity over complexity ❌✔


In summary, your tests could use a little refactoring but most of your methods are perfect.
The one exception is `Node.ChildItems` , this method generates an infinite loop!

File: Assignment/Assignment/SampleData.cs [hunk 5]
```diff
@@ -1,30 +1,79 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
 
 namespace Assignment
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:7]: A general consideration, when returning an `IEnumerable`, you don't need to use `ToList` to convert your return value to a list.

File: Assignment/Assignment.Tests/NodeTests.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+
+namespace GenericsHomework.Tests
+{
+    [TestClass]
+    public class NodeTests
```
Comment (mmwoodfo) [inline Assignment/Assignment.Tests/NodeTests.cs:7]: You could use tests in here for your IEnumerable implementation

File: Assignment/Assignment.Tests/SampleDataTest.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            SampleData sampleData = new();
+
+            string csvRowsResult = sampleData.GetAggregateSortedListOfStatesUsingCsvRows();
+            string givenPeopleResult = sampleData.GetAggregateListOfStatesGivenPeopleCollection(sampleData.People);
+
+            Assert.AreEqual<string>(csvRowsResult, givenPeopleResult);
+        }
+
+        //private IEnumerable<Address> GetSpokaneAddresses()
```
Comment (mmwoodfo) [inline Assignment/Assignment.Tests/SampleDataTest.cs:86]: You should have a test that uses a hard-coded list of Spokane addresses.
In general, you could use more tests here.

**Review by mmwoodfo — COMMENTED** [review]
I think you have some good PR comments that should have been addressed.

**Comment by mmwoodfo** [comment]
@dhoxie Did you do a PR review?
